### ISSUE TITLE
Can't fold split window, if folds not ready before splitting

### ISSUE THREAD

**Issue by skt041959** [kevinhwang91/nvim-ufo#78]
### Neovim version (nvim -v | head -n1)

NVIM 0.7.2

### Operating system/version

Linux

### How to reproduce the issue

`cat mini.lua`
```lua
-- Use Vim packages install the plugin, also work with some plugins manager such as packer.nvim
vim.o.packpath = '~/.local/share/nvim/site'
vim.cmd('packadd promise-async')
vim.cmd('packadd nvim-ufo')

-- Setting
vim.o.foldcolumn = '1'
vim.o.foldlevel = 99
vim.o.foldlevelstart = -1
vim.o.foldenable = true

local ufo = require('ufo')
ufo.setup()
vim.keymap.set('n', 'zR', ufo.openAllFolds)
vim.keymap.set('n', 'zM', ufo.closeAllFolds)
```
`nvim --clean +'so mini.lua'`

1. `:edit test.cpp`
2. `<C-w>v` (should action before folds get created)
3. `zM`
4. `<C-w>l`
5. `zM`

### Expected behavior

Both windows can be folded

### Actual behavior

The second window don't fold after `zM`

**Comment by kevinhwang91** [comment]
Can't reproduce. Two windows are folded as expected.

**Comment by skt041959** [comment]
> Can't reproduce. Two windows are folded as expected.

I guess you did step 2 too late, otherwise your machine is too fast.
You need to split the window, before it shows the fold column.

**Comment by kevinhwang91** [comment]
Sorry, I omitted the content in the bracket in step2. The behavior is expected for now. The data is based on the buffer, but applying folds is based on the window.

It's hard to fix the issue for ufo. If you have an idea to handle the logic, feel free to share it here.

**Comment by skt041959** [comment]
I think an event emit for `WinEnter` could be enough.

**Comment by skt041959** [comment]
> I think an event emit for `WinEnter` could be enough.

Actually, I don't think we even need autocmd. I just need it to create fold every time I press `zM`.

**Comment by kevinhwang91** [comment]
ufo get window by using `bufwinid(bufnr)`. If the two windows are with the same buffer and in the second window currently, `bufwinid` will give you the first window. And if we change the buffer in second window, and the first window will apply fold.

The real issue is keeping folds consistent for all windows with the same buffer.

**Comment by skt041959** [comment]
I understand that the API `bufwinid` only returns the first window displaying that buffer. But I don't think it should create folds for all windows with that buffer or keep consistent.

The really important thing is, when I press `zM`, I need it act. I don't mind it calculate or query folds from language server again.

**Comment by kevinhwang91** [comment]
> The really important thing is, when I press zM, I need it act. I don't mind it calculate or query folds from language server again.

Make sense, I will try to fix it.

**Comment by kevinhwang91** [comment]
1. If the window in 1st step has applied folds, the second window created by `vsp` will inherit the folds.
2. If the window in 1st step hasn't applied folds, fire `InsertLeave` or `BufWritePost` event will reapply folds.

The second condition is not perfect, but I don't think it's worth spending too much time handling this edge case.

**Comment by skt041959** [comment]
@kevinhwang91 Seems the issue in OP still exists.

**Comment by kevinhwang91** [comment]
> The second condition is not perfect, but I don't think it's worth spending too much time handling this edge case.

Type `i<Esc>`

**Comment by skt041959** [comment]
> > The second condition is not perfect, but I don't think it's worth spending too much time handling this edge case.
> 
> Type `i<Esc>`
Could `zM` do the same work?

**Comment by skt041959** [comment]
I mean, could the five steps in OP just work without any other steps to bear in mind?

**Comment by kevinhwang91** [comment]
> I mean, could the five steps in OP just work without any other steps to bear in mind?

No, depending on the state of 2nd step. This is an edge case for your workflow. ufo just make multiple windows with the same buffer as normal as possible (Although it is impossible to be perfect). You should use `WinEnter` to check out whether the windows in the current tabpage with same buffers and do your post action.

**Comment by kevinhwang91** [comment]
`open_fold_hl_timeout` work fine now.

**Comment by skt041959** [comment]
>  You should use `WinEnter` to check out whether the windows in the current tabpage with same buffers and do your post action.

Thanks for your reply. 

To be honest, I don't think it's an edge case. Let me make it clear.
In my opinion, `closeAllFolds()` should always check whether Ufo could but didn't create folds in this window.
For another case(**not tested though**), when `closeAllFolds()` is called in a window where the folds are created from indent(fallback), but the language server provider is ready, Ufo should refresh folds from languages server.

To sum up, I think I'm asking for a force-refresh method. Why does it need to manually fire an additional `InsertLeave` event?
